cmd/cloud-endpoints-controller/main.go

```diff
@@ -364,7 +392,34 @@ func calcParentSig(parent *CloudEndpoint, addStr string) string {
 	return fmt.Sprintf("%x", hasher.Sum(nil))
 }
 
-func makeDefaultOpenAPISpec(endpoint, target string, jwtAudiences []string) (string, error) {
+func validateOpenAPISpec(specOriginal string) error {
+	var spec map[string]interface{}
+	return yaml.Unmarshal([]byte(specOriginal), &spec)
+}
+
+func getUserSuppliedOpenAPISpec(parent *CloudEndpoint, status *CloudEndpointControllerStatus) (string, error) {
+	var specString string
+	var err error
+	if parent.Spec.OpenAPISpec != nil {
+		specYaml, err := yaml.Marshal(parent.Spec.OpenAPISpec)
+		if err != nil {
+			log.Printf("[ERROR][%s] Supplied spec: %s", parent.Name, err.Error())
+			return "", errors.Wrap(err, "Could not load user supplied API from CloudEndpoint Spec")
+		}
+		specString = string(specYaml)
+	} else if name, key := parent.Spec.OpenAPISpecConfigMap.Name, parent.Spec.OpenAPISpecConfigMap.Key; name != "" && key != "" {
+		specString, err = getConfigMapSpecData(parent.ObjectMeta.Namespace, name, key)
+		if err != nil {
```

Comment:
This needs to wait gracefully for the configmap to become available. Similar to how it waits for the a target ingress to become available. Think about the cases where the cloudep is created before the configmap, the controller shouldn't go into a hard fail loop as it knows it should be waiting on the configmap.

Reply:
I think it will in the case of err != nil the status gets returned.  I need to test it though.  However, I think there may be an issue with the getTargetIngress() call.